# Combined Data Analysis with `get_combined_data.sh` and `run_PCA.sh`

This folder contains a set of scripts used for advanced analyses, including combining outputs from the SAUNA peak caller and generating various plots related to nucleosome positioning and gene expression.

## Combined Data Analysis with `get_combined_data.sh`

The main script in this folder is `get_combined_data.sh`, which performs the following analyses:

- Combines the outputs from the SAUNA peak caller for each sample.
- Calculates interpeak distances between nucleosomes.
- Plots frequencies of interpeak distances across AB subcompartments.
- Plots frequencies of interpeak distances across ChromHMM compartments.
- Generates a plot for summed fragment centers around transcription start sites (TSS) of the 2000 most highly expressed genes.
- Generates a smoothed peak count plot for positions around the TSS of the 2000 most highly expressed genes.

### Usage

The `get_combined_data.sh` script is run with the following command:
./get_combined_data.sh <bam_directory> <AB_compartment_annotation_file> <ChromHMM_compartment_annotation_file> <gene_expression_file> <TSS_annotation_file>


- `<bam_directory>`: Directory containing BAM files and associated subdirectories (organized as per the `run_everything.sh` script).
- `<AB_compartment_annotation_file>`: Annotation file for AB subcompartments.
- `<ChromHMM_compartment_annotation_file>`: Annotation file for ChromHMM compartments.
- `<gene_expression_file>`: File containing gene expression data.
- `<TSS_annotation_file>`: Annotation file for transcription start sites (TSS).

### File Organization

The `bam_directory` should follow the same organization as created by the `run_everything.sh` script found in the 'Running_peak_callers' folder. This structure ensures that:

- All BAM files are located in the `bam_directory`.
- For each BAM file, there is a corresponding folder containing the following subdirectories:
  - `input`: Contains `tsv.gz` files with fragment center data as well as `tsv` files generated by MaxFinder.
  - `output`: Contains `bed` files generated by the SAUNA peak caller.
  - `archive`: Used during the scriptâ€™s execution, but should be empty after the script completes.
  - `other`: Contains unused intermediate files after script execution.

### Input Requirements

The script requires the following inputs:

- **Bam Directory**: Containg TSV.GZ files with fragment center data and BED files with output generated by the SAUNA peak caller.
- **Annotation Files**:
  - **AB Subcompartments**: BED file annotating genomic regions with A and B subcompartments.
  - **ChromHMM**: BED file with ChromHMM annotations for various compartments.
  - **Gene Expression Data**: TSV.GZ file with gene expression values, required for generating TSS-related plots.
  - **TSS Annotation File**: BED file with TSS positions.

### Outputs

- **Interpeak Distance Plots**: Visualizations showing the distribution of interpeak distances across AB subcompartments and ChromHMM compartments. Can be found in the corresponding output directories for each sample.
- **TSS-Centered Plots**:
  - Summed fragment centers relative to the TSS of the 2000 most expressed genes. Can be found in the corresponding input directories for each sample.
  - Smoothed peak counts for positions +/- 1000 bp relative to TSS. Can be found in the corresponding output directories for each sample.

## PCA and Gene Enrichment Analysis with `run_PCA.sh`

The script `run_PCA.sh` is used to perform PCA (Principal Component Analysis) on interpeak distance data from SeCT samples. Additionally, it conducts a gene enrichment analysis based on bins where interpeak distances differ significantly between cancer and benign samples.

### Usage

To run the PCA and gene enrichment analysis, use the following command:
./run_PCA.sh <directory> <chromosome_lengths> <blacklisted_regions> <genes_file>


### Input Requirements

- **directory**: A directory containing one interpeak distance file for each SeCT sample (9 in total), named `SampleName_interpeak_distance.bedgraph`. Each file should be formatted as a bedgraph file with 5 columns, where the 5th column contains the interpeak distance values.
- **chromosome_lengths**: TXT file with the lengths of each chromosome, used for dividing the genome into 3000 bins.
- **blacklisted_regions**: BED file with genomic regions to exclude from the analysis.
- **genes_file**: TSV.GZ file with gene information for performing gene enrichment analysis.

### Analyses Performed

- **PCA**: Bins the genome into 3000 bins and calculates the mean interpeak distances for each bin and each SeCT sample. Performs PCA on the resulting data and plots the PCA results.
- **Gene Enrichment Analysis**: Identifies bins where interpeak distances differ significantly between high and low tumor fraction samples and performs gene enrichment analysis on these bins.Two TSV files are generated:
  1. **Cancer-lower.tsv**: Enrichment terms for bins where interpeak distances are shorter in high tumor fraction samples.
  2. **Benign-lower.tsv**: Enrichment terms for bins where interpeak distances are shorter in low tumor fraction samples.

### Outputs

- **PCA Plot**: Visualizes the PCA with PC1 and PC2 for the SeCT samples.
- **Gene Enrichment TSVs**: Two TSV files containing the gene enrichment analysis results.

